compiler/src/iree/compiler/Codegen/Common/FlattenMemRefs.cpp

```diff
@@ -164,185 +164,151 @@ static bool checkLayout(Value val) {
 }
 
 namespace {
-struct FlattenMemrefLoad : public OpRewritePattern<memref::LoadOp> {
-  using OpRewritePattern::OpRewritePattern;
 
-  LogicalResult matchAndRewrite(memref::LoadOp op,
+template <typename T>
+struct MemRefRewritePatternBase : public OpRewritePattern<T> {
+  using OpRewritePattern<T>::OpRewritePattern;
+  LogicalResult matchAndRewrite(T op,
                                 PatternRewriter &rewriter) const override {
-    Value memref = op.getMemref();
+    Value memref = getTargetMemref(op);
     if (!needFlattenning(memref) || !checkLayout(memref))
       return rewriter.notifyMatchFailure(op,
                                          "nothing to do or unsupported layout");
-
     auto &&[flatMemref, offset] = getFlattenMemrefAndOffset(
-        rewriter, op.getLoc(), memref, op.getIndices());
+        rewriter, op->getLoc(), memref, op.getIndices());
+    replaceOp(op, rewriter, flatMemref, offset);
+    return success();
+  }
+  virtual Value getTargetMemref(Operation *op) const = 0;
```

Comment:
1. This should probably take a `T`?
2. Instead of making these virtual functions, they could be free-standing templated functions? But I'm not sure I have a strong opinion there.

Reply:
1. You are right. it should take a `T`. updated.
2. those are one-off functions, they are likely different when `T` is different, so ... a virtual function is imho a better thing here ...